client/network-gossip/src/bridge.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::Network;
+use crate::state_machine::{ConsensusGossip, Validator, TopicNotification};
+
+use network::Context;
+use network::message::generic::ConsensusMessage;
+use network::{Event, config::Roles};
+
+use futures::{prelude::*, channel::mpsc, compat::Compat01As03};
+use libp2p::PeerId;
+use parking_lot::Mutex;
+use sr_primitives::{traits::{Block as BlockT, NumberFor}, ConsensusEngineId};
+use std::{borrow::Cow, sync::Arc, time::Duration};
+
+pub struct GossipEngine<B: BlockT> {
+	inner: Arc<Mutex<GossipEngineInner<B>>>,
+	engine_id: ConsensusEngineId,
+}
+
+struct GossipEngineInner<B: BlockT> {
+	state_machine: ConsensusGossip<B>,
+	context: Box<dyn Context<B> + Send>,
+}
+
+impl<B: BlockT> GossipEngine<B> {
+	/// Create a new instance.
+	pub fn new<N: Network + Send + Clone + 'static>(
+		network: N,
+		proto_name: impl Into<Cow<'static, [u8]>>,
+		engine_id: ConsensusEngineId,
+		validator: Arc<dyn Validator<B>>,
+	) -> Self where B: 'static {
+		let proto_name = proto_name.into();
+
+		let mut state_machine = ConsensusGossip::new();
+		let mut context = Box::new(ContextOverService {
+			network: network.clone(),
+			proto_name: proto_name.clone(),
+		});
+
+		network.register_notif_protocol(proto_name.clone(), engine_id, Vec::new());
+		state_machine.register_validator(&mut *context, engine_id, validator);
+
+		let inner = Arc::new(Mutex::new(GossipEngineInner {
+			state_machine,
+			context,
+		}));
+
+		let gossip_engine = GossipEngine {
+			inner: inner.clone(),
+			engine_id,
+		};
+
+		async_std::task::spawn({
+			let inner = Arc::downgrade(&inner);
+			async move {
+				loop {
+					async_std::task::sleep(Duration::from_millis(1100)).await;
+					if let Some(inner) = inner.upgrade() {
+						let mut inner = inner.lock();
+						let inner = &mut *inner;
+						inner.state_machine.tick(&mut *inner.context);
+					} else {
+						break;
+					}
+				}
+			}
+		});
+
+		async_std::task::spawn(async move {
+			let mut stream = Compat01As03::new(network.events_stream());
+			while let Some(Ok(event)) = stream.next().await {
```

Comment:
In case sending fails within network worker we drop the sender, thus the channel will be closed. I think we should either log the send failure in network worker, or log that the channel is closed here.

Network worker:

```rust
				Ok(Async::Ready(Some(BehaviourOut::Event(ev)))) => {
					self.events_streams.retain(|sender| sender.unbounded_send(ev.clone()).is_ok());
					return Ok(Async::Ready(Some(ev)));
				},
```

Reply:
To me that's a legitimate way to end the task.
If the `NetworkWorker` gets destroyed, then we want this background task to get closed as well.

Reply:
Ah, and with the above down and upgrade of the `Arc` we ensure to end the other spawned task as well?